api/src/main/java/com/spotify/metrics/core/MetricId.java

```diff
@@ -55,20 +57,32 @@
     private final int hash;
 
     public MetricId() {
-        this(null, EMPTY_TAGS);
+        this(null, EMPTY_TAGS, true);
     }
 
     public MetricId(String key) {
-        this(key, EMPTY_TAGS);
+        this(key, EMPTY_TAGS, true);
     }
 
     public MetricId(String key, Map<String, String> tags) {
-        this(key, Collections.unmodifiableSortedMap(new TreeMap<>(tags)));
+        this(key, new TreeMap<>(tags), false);
     }
 
+    /**
+     * This should never be used except for a single test case.
+     *
+     * @param key
+     * @param tags
+     */
+    @Deprecated
+    @VisibleForTesting
     MetricId(String key, SortedMap<String, String> tags) {
+        this(key, tags, false);
+    }
+
+    private MetricId(String key, SortedMap<String, String> tags, boolean isAlreadyUnmodifiable) {
         this.key = key;
-        this.tags = tags;
+        this.tags = isAlreadyUnmodifiable ? tags : Collections.unmodifiableSortedMap(tags);
```

Comment:
just curious, you couldn’t check this with instanceof? I can’t recall if the type that unmodifiableMap returns is public / a part of the API

Reply:
I investigated that, but it turns out those classes are not public, so it would be some ugly reflection mess. I decided this was cleaner. Since the method is private, we can be sure that all callers can already know if it is unmodifiable or not.

Reply:
makes sense 👍